### PR TITLE
Fix guarding issues w/ numpy

### PR DESCRIPTION
**Pull request by voznesenskym**
```
import torch
import numpy as np


def _inf_nan_preprocess(t, t_np):
    t_np = np.nan_to_num(t_np)
    return t, t_np


@torch.compile()
def fn():
    # shape, dims format
    test_cases = (
        (3, 3),
        (4, 4),
        (5, 5),
    )

    for shape in test_cases:
        t = torch.randn(shape, dtype=torch.complex64)
        t_np = np.random.randn(*shape).astype(np.complex64)

        _, t_np = _inf_nan_preprocess(t, t_np)
        print(t, t_np)  # Just a side effect so that compilation kicks in


fn()
```

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

File: torch/_dynamo/config.py
Comment (voznesenskym) [inline torch/_dynamo/config.py:101]: @lezcano this is intentional, this flag makes it hard to debug.

File: torch/_dynamo/guards.py
Comment (voznesenskym) [inline torch/_dynamo/guards.py:573]: if .get() invokes an fn, like in the source added in this PR (NumpyTensorSource) - we do not have id stability, which is required between example values from `get` and traced values at fake-ification time, because of how we build the sizes/strides for tensor guards.

**Review by lezcano — COMMENTED** [review]
Note that, with this PR in the current state, we still have that a function that just has NumPy inputs and NumPy outputs is not being compiled. A small modification of the example in the OP shows this.

OTOH, they seem to have correct guards, and their shapes are correctly traced with symints when dynamic shapes kick in.

File: torch/_dynamo/utils.py
Comment (lezcano) [inline torch/_dynamo/utils.py:1649]: Let's save a few roundtrips
```suggestion
        return torch.as_tensor(value)
```

File: torch/_dynamo/utils.py
Comment (lezcano) [inline torch/_dynamo/utils.py:1650]: nit. `elif`

File: torch/_dynamo/source.py
Comment (lezcano) [inline torch/_dynamo/source.py:487]: Perhaps we can simply call `torch.as_tensor` here?

File: torch/_dynamo/variables/tensor.py
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:491]: Is this just a refactor?

**Comment by voznesenskym** [comment]
@lezcano I undid your mega merge, this PR is back to being based on head of your branch. If you need work there, push it to that branch and rebase this branch, don't push to this branch please :P

File: torch/_dynamo/source.py
Comment (voznesenskym) [reply torch/_dynamo/source.py:487]: It should be consistent with whatever we used during tracing and guarding. So if we do, let's move everything in this impl to that.

File: torch/_dynamo/utils.py
Comment (voznesenskym) [reply torch/_dynamo/utils.py:1650]: the branches return, doesnt matter, but ok

File: torch/_dynamo/codegen.py
Comment (ezyang) [inline torch/_dynamo/codegen.py:160]: does this... do anything lol

File: torch/_dynamo/guards.py
Comment (ezyang) [inline torch/_dynamo/guards.py:575]: numpy ndarray on nn module 🤔  hilarious failure mode

File: torch/_dynamo/guards.py
Comment (ezyang) [reply torch/_dynamo/guards.py:573]: I think I really need a comment here explaining what the value arg does. Still not sure from the comment either, need to read more.

File: torch/_dynamo/utils.py
Comment (ezyang) [reply torch/_dynamo/utils.py:1649]: Can only do as_tensor if you get an ndarray. For other types you need different handling.

File: torch/_dynamo/variables/tensor.py
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:885]: What's going on here? Are there two proxies floating around now?

File: torch/_dynamo/variables/builder.py
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:981]: It appears these are the substantive changes

File: torch/_dynamo/variables/builder.py
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:168]: Who sets `__dynamo_retain`?

File: torch/_dynamo/variables/builder.py
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:936]: So are you passing value for EVERYBODY, not just ndarray?

File: torch/_dynamo/variables/builder.py
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:976]: I guess I don't really understand why you can't just pass this source directly into `wrap_fx_proxy_cls`, with the idea that it's going to create a guard for the tensor_value you pass it, but then disgorge a NumpyNdarrayVariable. Creating a throwaway tensor_vt and smuggling the guards off it smacks me the wrong way.

File: torch/_dynamo/variables/builder.py
Comment (lezcano) [inline torch/_dynamo/variables/builder.py:975]: Let's move everything to `as_tensor`. Note that that's what's being done here, and it's what we want to do really

**Comment by lezcano** [comment]
>@lezcano I undid your mega merge, this PR is back to being based on head of your branch. If you need work there, push it to that branch and rebase this branch, don't push to this branch please :P

OOPS, sorry. I started working on top of this PR to figure out what was going on in that model, but then the fixes would stand on their own, and I cherry-pick them into my branch, but I forgot to rebase your branch, sorry!

**Comment by lezcano** [comment]
Ok, keeping a nice commit history in the main PR via rebases and `push -f`s and keeping it on this one at the same time is a bit of a mess. I'll try not to force-push the main branch until this one is merged.

**Comment by lezcano** [comment]
The main PR has all the tests passing minus the one this one is fixing. Once this one is ready to be merged I think the CI in the main PR should be fully green.

**Comment by voznesenskym** [comment]
> The main PR has all the tests passing minus the one this one is fixing. Once this one is ready to be merged I think the CI in the main PR should be fully green.

Sounds good will get this over.

File: torch/_dynamo/codegen.py
Comment (voznesenskym) [reply torch/_dynamo/codegen.py:160]: Its the better convention, but no, should be same.

File: torch/_dynamo/guards.py
Comment (voznesenskym) [reply torch/_dynamo/guards.py:573]: okay, will add a comment.

File: torch/_dynamo/guards.py
Comment (voznesenskym) [reply torch/_dynamo/guards.py:575]: This will not fail, but will overspecialze.

File: torch/_dynamo/utils.py
Comment (voznesenskym) [reply torch/_dynamo/utils.py:1649]: Sg

File: torch/_dynamo/variables/builder.py
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:168]: Nobody, leftover

File: torch/_dynamo/variables/builder.py
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:950]: Yes, this actually saves us an eval, and should always be sound.

File: torch/_dynamo/variables/builder.py
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:990]: It's not great, but it keeps guard preservation 1:1 with the tensor. I was initially going to manually do this, but then I realized that if tensor guards diverge, this would be kind of a weak contract to enforce. This forces them to be the same.

File: torch/_dynamo/variables/tensor.py
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:885]: We probably don't need this line, super already does that

File: torch/_dynamo/variables/tensor.py
Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:491]: Yeah, It was a little tougher to debug w/o this.

File: torch/_dynamo/utils.py
Comment (lezcano) [inline torch/_dynamo/utils.py:1649]: Better to remove this one, as, if our logic is correct, an `np.ndarray` should never get to this point. We should probably even assert `not isinstance(value, np.ndarray)`

**Comment by lezcano** [comment]
The issue in https://github.com/pytorch/pytorch/pull/106431#pullrequestreview-1558515416 is still present (i.e., if you remove the tensor arg to the `_inf_nan_preprocess` function we don't compile anything) but this shouldn't be blocking to land the main PR, I don't think.

**Comment by voznesenskym** [comment]
> The issue in [#106431 (review)](https://github.com/pytorch/pytorch/pull/106431#pullrequestreview-1558515416) is still present (i.e., if you remove the tensor arg to the `_inf_nan_preprocess` function we don't compile anything) but this shouldn't be blocking to land the main PR, I don't think.

Let me take a look :)

**Comment by voznesenskym** [comment]
> > The issue in [#106431 (review)](https://github.com/pytorch/pytorch/pull/106431#pullrequestreview-1558515416) is still present (i.e., if you remove the tensor arg to the `_inf_nan_preprocess` function we don't compile anything) but this shouldn't be blocking to land the main PR, I don't think.
> 
> Let me take a look :)

BTW this would be in another PR - its not blocking this one. I have not looked yet, but my money is on something that looks for tensor compute in frames.

File: torch/_dynamo/utils.py
Comment (voznesenskym) [reply torch/_dynamo/utils.py:1649]: Let me double check, I am pretty certain we do hit this, and that is why I added it. Why would you expect us to not get here?

**Review by lezcano — APPROVED** [review]
Yep, another PR SGTM.

On the interaction with `torch_np`, this PR LGTM. Let's wait for Ed's review about the general strategy tho.

File: torch/_dynamo/variables/misc.py
Comment (larryliu0820) [inline torch/_dynamo/variables/misc.py:846]: What is the difference before and after? Or is this just refactoring.

**Comment by lezcano** [comment]
I merged it because @voznesenskym was going to. 
This has been merged into https://github.com/pytorch/pytorch/pull/106211. If any of the reviewers @yanboliang @anijain2305 @ezyang have any comments, you can put them there and I'll address them.

**Pull request merged**
